heka/sandbox/filters/dollars.lua

```diff
@@ -0,0 +1,268 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+-- Run tests using `lua -e "TestMode = true;" dollars.lua`
+
+--[[
+Monitor submission sizes and counts by channel and document type. This tracks
+factors relevant to budget planning. See Bug 1179751.
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [PipelineBudget]
+    type = "SandboxFilter"
+    filename = "lua_filters/dollars.lua"
+    message_matcher = "Type == 'telemetry'"
+    ticker_interval = 60
+    preserve_data = false
+
+--]]
+
+require "os"
+require "table"
+if not TestMode then
+    require "lpeg"
+    require "cjson"
+
+    fx = require "fx"
+    max_per_channel = read_config("max_per_channel") or 60
+    -- YYYYMMDD<whatever>
+    bdate_grammar = lpeg.R("09") ^ 8
+end
+
+-- Size and count by day
+local sizes = {
+    submission = {},
+    build = {},
+}
+
+-- Key count by channel (for eviction purposes)
+local counters = {
+    submission = {},
+    build = {},
+}
+
+function get_build_date(buildid)
+    if not buildid then
+        return buildid
+    end
+
+    if string.len(buildid) < 8 then
+        return nil
+    end
+
+    if not lpeg.match(bdate_grammar, buildid) then
+        return nil
+    end
+
+    return string.sub(buildid, 1, 8)
+end
+
+-- If possible, evict a key older than "newdate" to make room for the new key.
+function evict(counter, newdate)
+    local evictable_keys = {}
+    local count = 0
+    for k, v in pairs(counter) do
+        count = count + 1
+        if v ~= nil and k < newdate then
+            table.insert(evictable_keys, k)
+        end
+    end
+    if #evictable_keys == 0 then
+        return 0
+    end
+    table.sort(evictable_keys)
+    local to_delete = 0
+    while count >= max_per_channel do
+        -- wipe out the min key.
+        to_delete = to_delete + 1
+        counter[evictable_keys[to_delete]] = nil
+        count = count - 1
+    end
+    return to_delete
+end
+
+function record_size(sizes, counter, channel, date, docType, size, count)
+    -- By default, count 1.
+    if not count then
+        count = 1
+    end
+
+    if not sizes[channel] then
+        sizes[channel] = {}
+        counter[channel] = 0
+    end
+    local sc = sizes[channel]
+
+    if not sc[date] then
+        if counter[channel] >= max_per_channel then
+            removed = evict(sc, date)
+            if removed == 0 then
+                -- This date is too old and our table is full.
+                return
+            end
+            counter[channel] = counter[channel] - removed
+        end
+        counter[channel] = counter[channel] + 1
+        sc[date] = {}
+    end
+    local scd = sc[date]
```

Comment:
this could be assigned before the if (instead of always looking it up twice) and explicitly set to the new table in the failed case.